src/rules/no-untyped-mock-factory.ts

```diff
@@ -0,0 +1,95 @@
+import { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/utils';
+import { createRule, isTypeOfJestFnCall } from './utils';
+
+const hasReturnType = (
+  node:
+    | TSESTree.FunctionExpression
+    | TSESTree.ArrowFunctionExpression
+    | TSESTree.Node,
+): boolean => {
+  if (
+    node.type === AST_NODE_TYPES.FunctionExpression ||
+    node.type === AST_NODE_TYPES.ArrowFunctionExpression
+  ) {
+    return node.returnType !== undefined;
+  }
+
+  return false;
+};
+
+const findModuleName = (
+  node: TSESTree.Literal | TSESTree.Node,
+): TSESTree.StringLiteral | null => {
+  if (node.type === AST_NODE_TYPES.Literal && typeof node.value === 'string') {
+    return node;
+  }
+
+  return null;
+};
+
+export type MessageIds = 'addTypeParameterToModuleMock';
+
+const DEFAULT_MESSAGE =
+  'Add a type parameter to the mock factory such as `typeof import({{ moduleName }})`';
+
+export default createRule<[], MessageIds>({
+  defaultOptions: [],
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description:
+        'Disallow using `jest.mock()` factories without an explicit type parameter',
+      recommended: false,
+    },
+    messages: {
+      addTypeParameterToModuleMock: DEFAULT_MESSAGE,
+    },
+    fixable: 'code',
+    schema: [],
+    type: 'suggestion',
+  },
+  create(context) {
+    return {
+      CallExpression(node: TSESTree.CallExpression): void {
+        const { callee, typeParameters } = node;
+
+        if (
+          callee.type === AST_NODE_TYPES.MemberExpression &&
+          node.arguments.length === 2
+        ) {
+          const { property } = callee;
+
+          if (
+            isTypeOfJestFnCall(node, context, ['jest']) &&
+            property.type === AST_NODE_TYPES.Identifier &&
+            ['mock', 'doMock'].includes(property.name)
```

Comment:
While we don't technically require all of our rules to support square brackets, you should be able to do that here by using `isSupportedAccessor` and `getAccessorValue` utilities (don't forget to add a couple of tests!)

Something like this is what you want:

```suggestion
            isSupportedAccessor(property) &&
            ['mock', 'doMock'].includes(getAccessorValue(property))
```